### PR TITLE
fix: remove callbacks resulting in scales due to incomplete response

### LINKED ISSUE #2659: Pull-based autoscaling causes unrelated scale-ups due to incomplete GitHub response
### Checks

- [X] I've already read https://github.com/actions/actions-runner-controller/blob/master/TROUBLESHOOTING.md and I'm sure my issue is not covered in the troubleshooting guide.
- [X] I'm not using a custom entrypoint in my runner image

### Controller Version

0.27.4

### Helm Chart Version

0.23.3

### CertManager Version

1.21.1

### Deployment Method

Kustomize

### cert-manager installation

- Yes
- Yes

### Checks

- [X] This isn't a question or user support case (For Q&A and community support, go to [Discussions](https://github.com/actions/actions-runner-controller/discussions). It might also be a good idea to contract with any of contributors and maintainers if your business is so critical and therefore you need priority support
- [X] I've read [releasenotes](https://github.com/actions/actions-runner-controller/tree/master/docs/releasenotes) before submitting this issue and I'm sure it's not due to any recently-introduced backward-incompatible changes
- [X] My actions-runner-controller version (v0.x.y) does support the feature
- [X] I've already upgraded ARC (including the CRDs, see charts/actions-runner-controller/docs/UPGRADING.md for details) to the latest and it didn't fix the issue
- [X] I've migrated to the workflow job webhook event (if you using webhook driven scaling)

### Resource Definitions

```yaml
---
apiVersion: actions.summerwind.dev/v1alpha1
kind: RunnerDeployment
metadata:
  labels:
    app: gha-runner-controller
  name: aws-core-2-unstable
  namespace: gha-runner
spec:
  template:
    spec:
      env:
      - name: DISABLE_RUNNER_UPDATE
        value: 'true'
      ephemeral: true
      image: summerwind/actions-runner:v2.304.0-ubuntu-22.04
      labels:
      - aws-core-2-default
      nodeSelector:
        nodegroup: core-8-unstable
      organization: myorg
      resources:
        limits:
          cpu: '1.8'
          memory: 6Gi
        requests:
          cpu: '1.8'
          memory: 6Gi
      serviceAccountName: agent
      volumeMounts:
      - mountPath: /etc/arc/hooks/job-started.d/label-podself.sh
        name: actions-runner-agent-config
        subPath: label-podself.sh
      volumes:
      - configMap:
          defaultMode: 493
          name: actions-runner-agent-config
        name: actions-runner-agent-config
---
apiVersion: actions.summerwind.dev/v1alpha1
kind: HorizontalRunnerAutoscaler
metadata:
  labels:
    app: gha-runner-controller
  name: aws-core-2-unstable
  namespace: gha-runner
spec:
  maxReplicas: 20
  metrics:
  - repositoryNames:
    - repo1
    - repo2
    - repo3
    type: TotalNumberOfQueuedAndInProgressWorkflowRuns
  minReplicas: 0
  scaleDownDelaySecondsAfterScaleOut: 60
  scaleTargetRef:
    name: aws-core-2-unstable
```

### To Reproduce

```markdown
1. Use RunnerDeployment with HorizontalRunnerAutoscaler. Use a label that not a single workflow is using since we want to showcase an unrelated trigger.
2. Only use pull-based scaling by using `TotalNumberOfQueuedAndInProgressWorkflowRuns` and defining some repositories. The more active those repositories the better. Active repositories = lots of GHA action is important since the controller is checking every single run whether it's of relevance and this introduced the unwanted behaviour.
3. Wait for an unrelated scale-up to happen. This is triggered by the GitHub API. Either it's an edge case in their API or eventual consistency resulting in them returning an empty Job Array for a run that actually has jobs.
```

### Describe the bug

GitHub's API is not always returning valid responses, while those are 200, they're empty.
The issue happens in the [listWorkflowJobs](https://github.com/actions/actions-runner-controller/blob/master/controllers/actions.summerwind.net/autoscaling.go#LL122C2-L122C18) function.
Two edge cases are caught by checking whether the `runID == 0` or whether the GH API response contains no jobs `len(allJobs) == 0`. In both cases, a callback is executed, which doesn't make any sense.
Neither case should trigger the callback of increasing the `queued` or `in_progress` counter, which ultimately determines the suggested desired amount of runners.
Ultimately, this results in unrelated RunnerDeployments being scaled since any workflow request can return an empty job array.

### Describe the expected behavior

If the Job Array from the GitHub API response is empty, we should not blindly scale RunnerDeployments but rather ignore it and let the next pull scrape check again.

### Whole Controller Logs

```shell
I added my own debugging to figure out what was going on.
I changed the repo and job / run id. I checked them, they do contain jobs and they were some GitHub hosted label `ubuntu-`.


2023-06-08T12:59:12Z	INFO	horizontalrunnerautoscaler	Checking workflows for queued
2023-06-08T12:59:12Z	INFO	horizontalrunnerautoscaler	Found the following runnerLabels	{"runner_labels": ["aws-core-2-default"], "runner_name": "aws-core-2-unstable", "repository": "repo"}
2023-06-08T12:59:12Z	INFO	horizontalrunnerautoscaler	Triggered CallBack - AllJobs == 0	{"run_id": 111111111, "repository": "repo"}
2023-06-08T12:59:12Z	INFO	horizontalrunnerautoscaler	Checking workflows for queued
2023-06-08T12:59:12Z	INFO	horizontalrunnerautoscaler	Found the following runnerLabels	{"runner_labels": ["aws-core-2-default"], "runner_name": "aws-core-2-unstable", "repository": "repo"}
2023-06-08T12:59:12Z	INFO	horizontalrunnerautoscaler	Detected a job	{"labels": ["ubuntu-22.04"], "run_id": 111111111, "job_id": 111111111, "repository": "repo"}
2023-06-08T12:59:12Z	INFO	horizontalrunnerautoscaler	Checking workflows for queued
2023-06-08T12:59:12Z	INFO	horizontalrunnerautoscaler	Found the following runnerLabels	{"runner_labels": ["aws-core-2-default"], "runner_name": "aws-core-2-unstable", "repository": "repo"}
2023-06-08T12:59:13Z	INFO	horizontalrunnerautoscaler	Detected a job	{"labels": ["ubuntu-22.04"], "run_id": 111111111, "job_id": 2222222222, "repository": "repo"}
2023-06-08T12:59:13Z	INFO	horizontalrunnerautoscaler	Checking workflows for queued
2023-06-08T12:59:13Z	INFO	horizontalrunnerautoscaler	Found the following runnerLabels	{"runner_labels": ["aws-core-2-default"], "runner_name": "aws-core-2-unstable", "repository": "repo"}
2023-06-08T12:59:13Z	INFO	horizontalrunnerautoscaler	Triggered CallBack - AllJobs == 0	{"run_id": 111111111, "repository": "repo"}
2023-06-08T12:59:16Z	DEBUG	horizontalrunnerautoscaler	Suggested desired replicas of 2 by TotalNumberOfQueuedAndInProgressWorkflowRuns	{"workflow_runs_completed": 0, "workflow_runs_in_progress": 0, "workflow_runs_queued": 2, "workflow_runs_unknown": 0, "namespace": "gha-runner-camunda", "kind": "runnerdeployment", "name": "aws-core-2-unstable", "horizontal_runner_autoscaler": "aws-core-2-unstable"}
2023-06-08T12:59:16Z	DEBUG	horizontalrunnerautoscaler	Calculated desired replicas of 2	{"horizontalrunnerautoscaler": "gha-runner-camunda/aws-core-2-unstable", "suggested": 2, "reserved": 0, "min": 0, "max": 20}
```
```


### Whole Runner Pod Logs

```shell
Runner Pod Logs are irrelevant since it's about the controller.
```

### Additional Context

I found this [discussion](https://github.com/actions/actions-runner-controller/commit/a63860029a85ac35c2839c19d3bcf3b5294377e5#r67617091) about the callback stuff.
I checked the code and I don't see a reason for keeping the callback behaviour and would suggest for both edge cases of `allJobs == 0` and `runID == 0` to simply ignore them rather than scaling up RunnerDeployments.

### PR DESCRIPTION
**Pull request by Langleu**
Fixes https://github.com/actions/actions-runner-controller/issues/2659

It removes the callbacks in the `listWorkflowJobs` function, which is only used by 2 edge cases, which shouldn't scale the RunnerDeployment.
GitHub requests can return an empty job array for a run even though it contains jobs. This results in unpredicted scale-ups of unrelated RunnerDeployments since no labels are checked but just blindly scaled.

I can also rework it into a feature flag to maintain the same functionality, but I currently don't see the use case.
The logging may be helpful for users to figure out stuck jobs, e.g. when GitHub Actions has an internal error, workflows can get stuck and it may not be noticeable to users. Therefore, helps to cancel those jobs once and for all that would otherwise trigger the previous edge case.

### CODE REVIEW

**Comment by sergeykranga** [comment]
after digging through the code base trying to understand why we see random scale ups of runners that actually don't do anything, we came to the same issue described here (thanks for opening the pull request too)!

this is affecting us as well, causing extra cost for ec2 instances. it would be great to get some attention from maintainers here

**Comment by Langleu** [comment]
@mumoshu, @toast-gear, @nikola-jokic anyone up or a review?
Would be nice to get it fixed since it can unknowingly result in tremendous bills.

**Review by mumoshu — DISMISSED** [review]
Hi @Langleu! First of all, awesome job! I can admit that the removed logic was helpful only when there were a few jobs with different labels- that could be cases 2 or 3 years ago. Today, it might be worse than nothing as many folks started using self-hosted runners for relatively more complex use-cases.

LGTM. Let me merge this, and I will try to cut a new ARC patch version this week!

Thank you for your contribution!

**Comment by Langleu** [comment]
Hey @mumoshu, thanks for having a look at it!

Any good suggestions on how to fix the tests?
From what I can see the tests are completely built around `runID == 0` callback, which in reality should never happen.

I played around a bit with it and we could give each single workflow run an id and mock via `workflowJobs` the response.
This only works if one passes a label like `self-hosted` since otherwise they're ignored. Which makes sense honestly, as the GHA controller shouldn't trigger on jobs that don't contain the `self-hosted` label.

**Comment by Langleu** [comment]
@mumoshu added in https://github.com/actions/actions-runner-controller/pull/2671/commits/45c975bbbc7e0d0ce85af5ec89ea06d296eadad6 how I imagine the tests to be fixed atm.
Some were unconsciously relying on the `runID == 0` callback, which is the default in tests but not the reality.

Additionally added 2 small test cases to make sure that neither a hosted runner nor an empty Jobs array is scaling the runner.

Looking forward to getting it merged and having a new release! 🚀

File: controllers/actions.summerwind.net/autoscaling_test.go
Comment (Langleu) [inline controllers/actions.summerwind.net/autoscaling_test.go:587]: this one is supposedly fixed at 3.
Maybe the test framework is broken but I didn't see the `fixed` value having any effect.
It only worked prior due to the `callback` secretly bumping the `in_progress` value

Comment (mumoshu) [reply controllers/actions.summerwind.net/autoscaling_test.go:587]: Good catch! However, I think this makes this test case useless- How about adding `"id":123` or whatever to every `workflow_runs` items so that it still falls into the desired code path, allowing this `want` value to remain `3`?

Comment (Langleu) [reply controllers/actions.summerwind.net/autoscaling_test.go:587]: @mumoshu, looked a bit through the code but I don't think it makes any difference.
The fixed value influences the replicas of the `RunnerDeployment` and that ultimately is the `scaleTarget` but to calculate the desired amount of replicas needed, none of that information is used.
The workflows don't have a `self-hosted` label, which would result in the requirement of 3. Otherwise they're treated as 3 hosted runners and that means the desired replicas == minimum defined as they're not of importance.

All autoscaler tests are running against the `AutoscalingMetricTypeTotalNumberOfQueuedAndInProgressWorkflowRuns` metric and none against `AutoscalingMetricTypePercentageRunnersBusy`. For the later, I could imagine the `fixed`, and therefore `scaleTarget` to have an influence but since we only check for `inQueue / inProgress` the `fixed` value and replicas of the RunnerDeployment have 0 influence on the `desired` value.

Comment (Langleu) [reply controllers/actions.summerwind.net/autoscaling_test.go:587]: @mumoshu, I'd propose removing the `fixed` tests, since they're irrelevant for `AutoscalingMetricTypeTotalNumberOfQueuedAndInProgressWorkflowRuns` tests.
Are you in accord with the change?

Comment (Langleu) [reply controllers/actions.summerwind.net/autoscaling_test.go:587]: @mumoshu , just another ping since this is stalling a bit ...

Comment (maxdanilov) [reply controllers/actions.summerwind.net/autoscaling_test.go:587]: @mumoshu pinging as well, would be amazing to have this PR merged for the stability of the pull-based approach.

Comment (alexanderkranga) [reply controllers/actions.summerwind.net/autoscaling_test.go:587]: @mumoshu another ping, we would really love this to get merged and released. This change is very important as it will have a positive impact on the cost of running multiple clusters

Comment (alexanderkranga) [reply controllers/actions.summerwind.net/autoscaling_test.go:587]: @nikola-jokic @toast-gear hey guys, any chance you can provide some feedback here too and help to get this merged? this is a very important change and we will all benefit from it

Comment (nikola-jokic) [reply controllers/actions.summerwind.net/autoscaling_test.go:587]: Hey, sorry for the delay,

I can't help much here since I am responsible for the `actions.github.com` API group. However, I asked @mumoshu to do another round of reviews and he is the best one to talk to related to the `actions.summerwind.net` group :relaxed: 

While waiting for the review, @Langleu do you mind updating the branch? That can fix rest of the tests. I can do it myself if you agree but it would be better if you do it

Comment (Langleu) [reply controllers/actions.summerwind.net/autoscaling_test.go:587]: @nikola-jokic, done.

Comment (nikola-jokic) [reply controllers/actions.summerwind.net/autoscaling_test.go:587]: Thank you!

Comment (Langleu) [reply controllers/actions.summerwind.net/autoscaling_test.go:587]: lint looks like a timeout, works locally.
Same for the RunnerSet related tests since this change doesn't touch that logic and runs into timeout as well.

Comment (nikola-jokic) [reply controllers/actions.summerwind.net/autoscaling_test.go:587]: No worries, I'll re-run the tests. I know it is not affecting that part of the code

File: controllers/actions.summerwind.net/autoscaling_test.go
Comment (mumoshu) [inline controllers/actions.summerwind.net/autoscaling_test.go]:
```suggestion
		// 0 demanded due to zero runID, min at 2
```

Perhaps this is better to say!
`want` becomes 2 due to the min replicas set to 2, not due to the workflow runs statuses demanded 2 replicas.

File: controllers/actions.summerwind.net/autoscaling_test.go
Comment (mumoshu) [inline controllers/actions.summerwind.net/autoscaling_test.go]:
```suggestion
		// 0 demanded due to zero runID, min at 2
```

File: controllers/actions.summerwind.net/autoscaling_test.go
Comment (mumoshu) [inline controllers/actions.summerwind.net/autoscaling_test.go]:
```suggestion
		// 3 demanded, max at 3
```

You actually demanded 3 replicas via those workflow jobs.

**Review by mumoshu — APPROVED** [review]
Sorry for the delay! Apparently, I got confused only due to how tests are commented.
I've corrected some comments, and this looks good overall now.
Thanks for your patience and contribution @Langleu!
// I've been a bit out of bandwidth due to a lack of sponsorships.

File: controllers/actions.summerwind.net/autoscaling_test.go
Comment (mumoshu) [reply controllers/actions.summerwind.net/autoscaling_test.go:587]: Thanks for your assistance @nikola-jokic!

**Pull request merged**
